### PR TITLE
Add #2155: newheightmapgame command

### PR DESCRIPTION
**Pull request by Berbe**
Work inspired by the latest patch from old FlySpray bug tracker: https://bugs.openttd.org/task/2155/getfile/8568/Heightmap_Console_command.patch in #2155:
- Added `newheightmapgame` command accepting up to 2 parameters (file name, title or index to load - mandatory, and an optional seed)
- The short version with no argument reloads the current heightmap with a random seed
Sadly, there is no clean way to have a 2-arguments version where the second one would be the seed, as it would conflict with the 2-arguments version where the integer represent the file ID. Even if file ID was to be discarded, there would be some complicated and dangerous checks to check if the 2nd parameter is a string or an integer that could potentially introduce vulnerabilities.

### CODE REVIEW

File: src/console_cmds.cpp [hunk 1]
```diff
@@ -1049,6 +1049,52 @@ DEF_CONSOLE_CMD(ConNewGame)
...
+
+		if (argc == 1) {
+			file = _file_to_saveload.title;
+		}
+
+		if (argc >= 2) {
+			file = argv[1];
+		}
+
+//		_console_file_list.ValidateFileList();
```
Comment (planetmaker) [inline src/console_cmds.cpp]: This commented-out line needs to go. Or should it not be commented-out?

File: src/genworld_gui.cpp [hunk 2]
```diff
@@ -878,6 +878,24 @@ void StartNewGameWithoutGUI(uint32 seed)
...
+ */
+void StartNewHeightMapGameWithoutGUI(uint32 seed)
+{
+	/* GenerateWorld takes care of the possible GENERATE_NEW_SEED value in 'seed' */
+	_settings_newgame.game_creation.generation_seed = seed;
+
+	uint x = 0;
+	uint y = 0;
+
+	/* If the function returns negative, it means there was a problem loading the heightmap */
```
Comment (planetmaker) [inline src/genworld_gui.cpp]: Comment says <0. Code says != 0.

**Review by planetmaker — CHANGES_REQUESTED** [review]
The commit message also doesn't follow style. Should go in the direction of
Add #2155: newheightmapgame command

File: src/console_cmds.cpp (hunk 1 above)
Comment (Berbe) [reply src/console_cmds.cpp]: Nah it shall disappear, as the ValidateFileList, despite its name, is oriented towards a specific filetype, as in normal savegames.
I wanted to pinpoint the oddity in this function behaviour.

That's why I have to manually invoke `BuildFileList(FT_HEIGHTMAP, SLO_LOAD);`

Comment (planetmaker) [reply src/console_cmds.cpp]: Maybe it warrants a comment in that direction?

File: src/genworld_gui.cpp (hunk 2 above)
Comment (Berbe) [reply src/genworld_gui.cpp]: Comment inherited from original patch, actually.
I guess the author, by 'negative', meant the boolean meaning as in 'not true', not the integer one.

I'll rephrase that

File: src/console_cmds.cpp (hunk 1 above)
Comment (Berbe) [reply src/console_cmds.cpp]: Wilco.

**Review by planetmaker — DISMISSED** [review]

File: src/console_cmds.cpp [hunk 3]
```diff
@@ -1049,6 +1049,54 @@ DEF_CONSOLE_CMD(ConNewGame)
...
+DEF_CONSOLE_CMD(ConNewHeightMapGame)
+{
+	bool heightmap_loaded = true;
+
+	if (argc == 0) {
+		IConsoleHelp("Load a game by name or index. Usage: 'newheightmapgame [<file | number> [seed]]'");
+		IConsoleHelp("The server can force a new game using 'newheightmapgame'; any client joined will rejoin after the server is done generating the new game.");
+	}
+
+	if (argc >=1 && argc <= 3) {
```
Comment (PeterN) [inline src/console_cmds.cpp]: Missing spacing. This could be changed to an early exit if argc < 1 || argc > 3, reducing the indentation depth below.

File: src/console_cmds.cpp [hunk 4]
```diff
@@ -1049,6 +1049,54 @@ DEF_CONSOLE_CMD(ConNewGame)
...
+				IConsolePrintF(CC_ERROR, "%s: Not a heightmap.", file);
+			}
+		} else {
+			IConsolePrintF(CC_ERROR, "%s: No such file or directory.", file);
+		}
+
+		_console_file_list.InvalidateFileList();
+	}
+
+	if (argc > 3) heightmap_loaded = false;
```
Comment (PeterN) [inline src/console_cmds.cpp]: Why is this handled all the way down here?

File: src/console_cmds.cpp [hunk 5]
```diff
@@ -1049,6 +1049,54 @@ DEF_CONSOLE_CMD(ConNewGame)
...
+
+	return heightmap_loaded;
```
Comment (PeterN) [inline src/console_cmds.cpp:1101]: If argc is 0 then heightmap_loaded will be true, although it didn't do it?

**Review by PeterN — DISMISSED** [review]

**Comment by Eddi-z** [comment]
the name sounds a bit clumsy, maybe `newgame_heightmap` would be a little clearer? not sure how that relates to other existing commands.

File: src/console_cmds.cpp (hunk 5 above)
Comment (Berbe) [reply src/console_cmds.cpp:1101]: Look at `ConNewGame`, it does exactly the same: explicitly returning `true` for that condition. That surprised me but I do not know better.

Are `IConsoleHelp` calls enough? Is returning `true` without doing anything else is *not* bad?
I am not enough of a pro to decide what should be, rather than what is, thus I merely mimic what is existing.

Original patch seems to have copied the `ConNewGame` section, thus reproducing this behaviour I did not change, merely rewrote to avoid multiple `return` scattered around.

File: src/console_cmds.cpp (hunk 4 above)
Comment (Berbe) [reply src/console_cmds.cpp]: I was sorting blocks by increasing number of arguments, considering this case would be less frequent than normal ones, thus concluding it would avoid an usually unecessary extra check for those cases, making the program perform better on average.

File: src/console_cmds.cpp (hunk 3 above)
Comment (Berbe) [reply src/console_cmds.cpp]: That leads to your question regarding returning `true` on `argc == 0`, and if that is good or bad. Returning `false` for `argc < 1` would be inconsistent with the programming of existing commands of major importance (to my eyes) like `ConNewGame`.
Please refer to aforementioned discussion to sort that out.

File: src/console_cmds.cpp (hunk 4 above)
Comment (PeterN) [reply src/console_cmds.cpp]: Seriously premature optimization there...

Comment (Berbe) [reply src/console_cmds.cpp]: Well I do not mind moving that up at all, judgments set aside. Where would that be, exactly?

**Comment by Berbe** [comment]
> the name sounds a bit clumsy, maybe `newgame_heightmap` would be a little clearer? not sure how that relates to other existing commands.

There is a `newgame` command. Out of lack of creativity, I merely reused the `newheightmapgame` command name as suggested in the original patch.
How do we decide what's best? :wink:

File: src/console_cmds.cpp (hunk 3 above)
Comment (nielsmh) [reply src/console_cmds.cpp]: `argc == 0` has a very specific meaning, which is "the user requested help on the command". It should always return true, and print the help message, for that case.

Comment (Berbe) [reply src/console_cmds.cpp]: Well that means the 'early exit' style would not trigger on `argc <1 || argc > 3`, but on `argc > 3`, like what I already did.
The question about using an 'early exit' style or merely setting the return variable value for a single `return` instruction shall be part of discussion around line 1095 below.

File: src/console_cmds.cpp (hunk 5 above)
Comment (nielsmh) [reply src/console_cmds.cpp:1101]: You should always early return after handling the help case at the top of the function, to make it super obvious that requesting help has no side effects.

Comment (nielsmh) [reply src/console_cmds.cpp:1101]: I believe the preferred style is to use a shallow control flow with early returns to reject invalid parameters, collect the valid parameters into a set of variables, and then execute the command on the top level of the function.

Note that if a command handler returns false, the console system will call the handler again with argc = 0 so the user gets a help string printed.

Comment (PeterN) [reply src/console_cmds.cpp:1101]: @berbe ConNewGame has a `return true;` inside the help block. You should have the same, please. This early return simplifies the code.

Comment (Berbe) [reply src/console_cmds.cpp:1101]: Done.

File: src/console_cmds.cpp [hunk 6]
```diff
@@ -1049,6 +1049,55 @@ DEF_CONSOLE_CMD(ConNewGame)
...
+{
+	if (argc == 0) {
+		IConsoleHelp("Load a game by name or index. Usage: 'newheightmapgame [<file | number> [seed]]'");
+		IConsoleHelp("The server can force a new game using 'newheightmapgame'; any client joined will rejoin after the server is done generating the new game.");
+		return true;
+	}
+
+	bool heightmap_loaded = true;
+
+	if (argc >=1 && argc <= 3) {
```
Comment (PeterN) [inline src/console_cmds.cpp]: argc >=1 is redundant, and the argc <= 3 should be as well.

File: src/console_cmds.cpp (hunk 4 above)
Comment (PeterN) [reply src/console_cmds.cpp]: Move it before the `if (argc >=1 && argc <= 3) {` test, and make it return early. That means the rest of the function doesn't need to be in an if condition and can be unindented a bit.

Comment (Berbe) [reply src/console_cmds.cpp]: Done.

File: src/console_cmds.cpp (hunk 6 above)
Comment (Berbe) [reply src/console_cmds.cpp]: This condition has vanished as part of your advice in another review.

**Comment by Berbe** [comment]
Added a new feature commit to allow CLI loading of heightmaps by filename/name/title

File: src/openttd.cpp [hunk 7]
```diff
@@ -622,19 +622,76 @@ int openttd_main(int argc, char *argv[])
...
 
-				/* if the file doesn't exist or it is not a valid savegame, let the saveload code show an error */
+				/* If supplied file does not exist or is not a valid savegame/scenario/heightmap, let the saveload code show an error */
 				const char *t = strrchr(_file_to_saveload.name, '.');
 				if (t != NULL) {
-					FiosType ft = FiosGetSavegameListCallback(SLO_LOAD, _file_to_saveload.name, t, NULL, NULL);
-					if (ft != FIOS_TYPE_INVALID) _file_to_saveload.SetMode(ft);
+					FiosType ft;
+					ft = FiosGetSavegameListCallback(SLO_LOAD, _file_to_saveload.name, t, NULL, NULL);
+					if (ft != FIOS_TYPE_INVALID) {_file_to_saveload.SetMode(ft); _switch_mode = SM_LOAD_GAME; break;}
```
Comment (glx22) [inline src/openttd.cpp]: You discard the is_scenario/is_heightmap stuff here, same for the next lines

Comment (Berbe) [reply src/openttd.cpp]: Those checks have been redone, along with those in blocks searching by name/title underneath. Are they better?

I seem however enable to start the editor by loading a heightmap there, irrespective of the loading method... I see a loading bar, then I get back to the game menu without any error message.
The question is: Is/Can my portion of code (be) responsible for that?

**Comment by glx22** [comment]
Made some suggestions in https://github.com/Berbe/OpenTTD/pull/1

**Review by glx22 — COMMENTED** [review]
I think StartNewHeightMapGameWithoutGUI() should return a bool, then you can use it in VideoDriver_Dedicated::MainLoop() to do a check similar to the savegame load check

**Comment by Berbe** [comment]
> 
> 
> I think StartNewHeightMapGameWithoutGUI() should return a bool, then you can use it in VideoDriver_Dedicated::MainLoop() to do a check similar to the savegame load check

That would kill the resemblance to `StartNewGameWithoutGUI()` which has a procedure signature.
If a check needs to be done, I guess we should rather modify/replicate `SafeLoad` to handle heightmaps.

**Comment by glx22** [comment]
Something like https://github.com/glx22/OpenTTD/commit/7c3adf34a89194db95b84897c01ae4ef743e70a9 should do it (modulo error message and comments)

**Comment by Berbe** [comment]
The problems I see with your solutions are multiple:
1.`StartNewHeightMapGameWithoutGUI` signature has to be changed from `void` to `bool`, drifiting away from `StartNewGameWithoutGUI`
2. On success, using that function to check will actually also generate a new map, which makes the game generate the map twice in total
To prevent that, further modifying the signature to add some `bool check` parameter would be required, drifting further

I chose another way: including specialized `heightmap.h` functions in `dedicated_v.cpp` and calling directly the already existing `GetHeightmapDimensions` function which checks for file validity (as it is already used for at https://github.com/Berbe/OpenTTD/blob/heightmap/src/genworld_gui.cpp#L836).
I do not know the side-effects of that inclusion, nor if it is wished, but it looks overall cleaner.

I'll let reviewers be the judge of that.

**Comment by nielsmh** [comment]
Looks like the change to use nullptr is the cause of conflicts.

Could this perhape be changed to instead extend the existing newgame command such that the second parameter can be a `-h` flag, instead of adding a new command?

I.e.:
`newgame scenarioname`
`newgame -h heightmapname`

And perhaps also allow `newgame -s scenarioname` for consistency.

**Comment by Berbe** [comment]
Yet another conflict? I rebased 4 days ago to remove them already... *sigh*
I will see to it.

**********

As for you suggestion of using flags/options, it is unheard of in console commands as far as I know: https://wiki.openttd.org/Console_Commands
'console' might be confusing, but we are not talking command-line, rather in-game console commands which follow far simpler syntax.

If there are global design decisions voted as part of a 'grand plan', I would be then gladly following them. Unless I am wrong, so far all I saw were individual decisions here and there, which IMHO hurt a *community* project. No command showing sign of options, I am concluding it has not been the followed design so far.
To avoid adding to the entropy, diverging from what actually exist, I will so far stick with the current way console commands are, anyone being free sharing an implementation of a more complex system.

File: src/openttd.cpp [hunk 8]
```diff
@@ -612,19 +612,84 @@ int openttd_main(int argc, char *argv[])
...
 				_file_to_saveload.SetName(mgo.opt);
-				bool is_scenario = _switch_mode == SM_EDITOR || _switch_mode == SM_LOAD_SCENARIO;
-				_switch_mode = is_scenario ? SM_LOAD_SCENARIO : SM_LOAD_GAME;
-				_file_to_saveload.SetMode(SLO_LOAD, is_scenario ? FT_SCENARIO : FT_SAVEGAME, DFT_GAME_FILE);
+				bool is_editor = _switch_mode == SM_EDITOR || _switch_mode == SM_LOAD_SCENARIO || _switch_mode == SM_LOAD_HEIGHTMAP;
+				_switch_mode = is_editor ? (_switch_mode == SM_LOAD_HEIGHTMAP ? SM_LOAD_HEIGHTMAP : SM_LOAD_SCENARIO) : SM_LOAD_GAME;
+
+				/* Unfortunately, initializing file type to being invalid breaks the SafeLoad function which expects SLO_LOAD operation: initializing to dummy value */
+				/*_file_to_saveload.SetMode(FIOS_TYPE_INVALID);
+				_file_to_saveload.SetMode(SLO_INVALID, FT_INVALID, DFT_INVALID);*/
```
Comment (LordAro) [inline src/openttd.cpp:623]: This commented out code should be removed

File: src/openttd.cpp [hunk 9]
```diff
@@ -612,19 +612,84 @@ int openttd_main(int argc, char *argv[])
...
+					}
+					ft = FiosGetHeightmapListCallback(SLO_LOAD, _file_to_saveload.name, t, nullptr, nullptr);
+					if (ft != FIOS_TYPE_INVALID) {
+						_file_to_saveload.SetMode(ft);
+						_switch_mode = is_editor ? SM_LOAD_HEIGHTMAP : SM_START_HEIGHTMAP;
+						break;
+					}
+				}
+
+				/* If supplied value was not recognized as a valid resource, attempt to find a matching title */
```
Comment (LordAro) [inline src/openttd.cpp:650]: I don't think loading by title is a good idea, especially not from the commandline (could result in confusing behaviour). Just filename will do, IMO

**Review by LordAro — CHANGES_REQUESTED** [review]

File: src/openttd.cpp (hunk 8 above)
Comment (Berbe) [reply src/openttd.cpp:623]: This comment underlines a problem which might be past my understanding.
Is the dummy value OK?
Shall I use `FIOS_TYPE_INVALID` instead? If so, is there a problem with `SafeLoad`? If not, what am I doing wrong?

File: src/openttd.cpp (hunk 9 above)
Comment (Berbe) [reply src/openttd.cpp:650]: Is there a consensus on that?
I find this feature quite handy & nice

Comment (TrueBrain) [reply src/openttd.cpp:650]: This PR grew a bit big :D Making it very difficult to review, as several different conversations go through each other now. This is a good example of that.

CLI tools should just use filenames; that is what users of CLI are used to, and going against that is only asking for problems. I am also not sure what issue it should fix to do it like that.

But as mentioned, if you want to have this in OpenTTD, please make it another PR, as having a conversation in a conversation makes for poor conversations :D

Tnx!

**Comment by TrueBrain** [comment]
I spend the last 4 hours or so going up and down the rabbithole that is called our console code. I found that it is difficult, and I think this PR is not the right approach.

Mostly, the issue is that `ls` only lists savegames. So I am not really sure how this was suppose to work, but when I do `newheightmapgame 1`, I have no idea which heightmap it is about to load :D So after some staring at the code, and trying various of things, I came to the conclusion there isn't really a way that is clean and clear to do this.

Additionally, we know for a while that `ls` and `cd` in the console are just a bad idea. So I think we should just change how you can list savegames etc, like: `list savegames` and `list scenarios` or something, and drop the navigating of files. Basically saying: if a savegame is not in the search path .. well .. maybe we shouldn't allow loading it from the console. But that is a bit outside this PR, but required for the PR in order the work.

Next, I looked at the command itself, and found out that I would expect `load` could also load scenarios, and possibly heightmaps. The latter is always a bit difficult, as you start a new game based on a heightmap .. do you really load it, is the question. But okay, semantics left there, this is something we can perfectly explain in the help, If people really object to that, I agree with a comment made in this PR: use `newgame -h <id>` or something. That would be much more intuitive.

Lastly, this PR had a bit of a feature-creep, where it also changed how `-g` works. I am very much against that change, for the simple reason that if you want `-g` to also work with "friendly" names, you should also allow a way to list them via the CLI. And all that code, I have to ask, what does it add? How much of a problem is using the filename over a "friendly" name in that case. It would only make sense if it was tight to the content service and could also download it, or something. Either way, this is just too much for this PR.

So in conclusion, and I do realise this is almost 2 years after the PR was created, sorry about that, I am going to close this PR. First someone needs to figure out what we want with the console, and what we want with file listing, and we can work up from there to support loading heightmaps and scenarios next to the savegames.

tl;dr: this PR does too many things, and the `newheightmapgame` command is in a practical sense not usable as presented in this PR. We need to go back a bit and rework how the console works for this to make sense.

**Comment by Berbe** [comment]
I am... speechless.

Being no developer, this PR (amongst the only few I made) was the most advanced one, feature-wise. It would not be an exaggeration to state that I was somewhat proud about it.

From a practical point of view, it all started because I wanted to improved the dedicated server version of the game, making it able to load heightmap games (a feature which is, to this day, still lacking), although the heightmap format is very powerful, practical and existing in the game for a long time now. It felt wrong it was unusable in dedicated servers.
Also from a practical point of view, I am using the code I pushed to you on my own dedicated server instances, which have been running flawlessly for a long time.

I understand the feature creep feeling, and I would have been more than eager to work with you on that, especially on the request to split this work in multiple PR if you saw that fit.

I understand the whole command-line thingy needs to be revamped, but IMHO it is no reason to reject attempts at iterating over it. Nothing can't be refactored later to follow yet-to-exist clear specifications.

I understand the code is a bit messy and convoluted, but again I did my best on the cadebase as it was. Again, I am no developer: if understanding the codebase was difficult for you, imagine what it was to me... and after much effort I was able to reach some working state, step by step.

I also understand the work I produced might not be of the highest quality standard, and maybe I myself introduced some off behaviour due to lack of skills and/or understanding of the codebase, but this is not an impossible issue to address.

To each problem its adaquate tackle. Piling on multiple problems (most of them beyond my reach) as a reason to send that contribution attempt away in a snap feels like being unfair.

I will continue running my modified servers, and I hope that, once the required groundwork to clean up the CLI interface is made, you will remember about this PR attempt and extract its main idea to implement it.

As of now, with that PR closed, all I see is a long-lacking feature which has been killed.
It is a bit overwhelming as, after 2 years, I am myself a bit rusty on all the details I put in those 3 orderly (or so I thought) commits.
I certainly did not expect it to end up being washed down the drain like that. In those 2 years, even with sporadic back-and-forth reviews and corrections, we could have achieved much. Instead, this eulogy followed by the trashbin. I am sorry we collectively spent that much time, involving that much people on this contribution attempts: this time is effectively lost for everyone.

Happy 2021, and so long.
